src/widget.c

```diff
@@ -211,32 +282,32 @@ extern void led_process_thread(void *d0, void *d1, void *d2) {
     while (true) {
         // wait until a blink item is received and process it
         struct blink_item blink;
+        uint8_t previous_color;
+        bool separation;
         k_msgq_get(&led_msgq, &blink, K_FOREVER);
-        LOG_DBG("Got a blink item from msgq, color %d, duration %d", blink.color,
-                blink.duration_ms);
+        if (blink.duration_ms > 0) {
+            LOG_DBG("Got a blink item from msgq, color %d, duration %d",
+                    blink.color, blink.duration_ms);
 
-        // turn appropriate LEDs on
-        for (uint8_t pos = 0; pos < 3; pos++) {
-            if (BIT(pos) & blink.color) {
-                led_on(led_dev, rgb_idx[pos]);
-            }
-        }
+            previous_color = led_current_color;
+            separation = blink.color == led_current_color && blink.color > 0;
 
-        // wait for blink duration
-        k_sleep(K_MSEC(blink.duration_ms));
-
-        // turn appropriate LEDs off
-        for (uint8_t pos = 0; pos < 3; pos++) {
-            if (BIT(pos) & blink.color) {
-                led_off(led_dev, rgb_idx[pos]);
+            if (separation) {
+                set_rgb_leds(0, CONFIG_RGBLED_WIDGET_INTERVAL_MS);
             }
-        }
+            set_rgb_leds(blink.color, blink.duration_ms);
+            if (separation) {
+                set_rgb_leds(0, CONFIG_RGBLED_WIDGET_INTERVAL_MS);
+            }
+            set_rgb_leds(previous_color, blink.sleep_ms);
```

Comment:
Looks good to me, thanks for the refactoring! One final nit, because it initially confused me trying to understand it and this seems equivalent (but easier to grok) to me, right?

```suggestion
            set_rgb_leds(previous_color, blink.sleep_ms > 0 ? blink.sleep_ms : CONFIG_RGBLED_WIDGET_INTERVAL_MS);
```
(I can't create a suggestion that includes lines below, but this would also remove lines 305-307, and we can move the comment on line 304 to before 302)

Reply:
Ah, yes, excellent idea! Thanks!